### PR TITLE
:star2: Add semantic segmentation trainer

### PR DESCRIPTION
**Pull request by edgarriba**
#### Changes

Fixes # (issue)

#### Type of change

- [x] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📝 This change requires a documentation update

#### Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/x/trainer_classification.py [hunk 1]
```diff
@@ -43,3 +45,34 @@ def evaluate(self) -> dict:
...
+class SemanticSegmentationTrainer(Trainer):
+    @torch.no_grad()
+    def evaluate(self) -> dict:
+        self.model.eval()
+        stats = {'losses': AverageMeter(), 'iou': AverageMeter()}
+        for sample_id, sample in enumerate(self.valid_dataloader):
+            sample = {"input": sample[0], "target": sample[1]}  # new dataset api will come like this
+            # perform the preprocess and augmentations in batch
+            sample = self.preprocess(sample)
+            sample = self.postprocess(sample)
```
Comment (shijianjian) [inline kornia/x/trainer_classification.py:59]: same here. why no augmentation?

File: kornia/contrib/classification.py [hunk 2]
```diff
@@ -24,3 +26,78 @@ def __init__(self, embed_size: int = 768, num_classes: int = 10) -> None:
     def forward(self, x: torch.Tensor) -> torch.Tensor:
         out = x.mean(-2)
         return self.linear(self.norm(out))
+
+
+def _expand(tensor, length: int):
+    return tensor.unsqueeze(1).repeat(1, int(length), 1, 1, 1).flatten(0, 1)
+
+
+class SegmentationHead(nn.Module):
```
Comment (shijianjian) [inline kornia/contrib/classification.py:35]: Why implement our own head than using torch vision models directly as in [here](https://github.com/pytorch/vision/blob/main/torchvision/models/segmentation/segmentation.py)? I would assume that we are implementing a trainer than modelling.

Comment (edgarriba) [reply kornia/contrib/classification.py:35]: There are some heads e.g. based on transformers that you cannot find yet

File: kornia/x/trainer.py [hunk 3]
```diff
@@ -95,19 +95,19 @@ def fit_epoch(self, epoch: int) -> None:
...
-            source, target = sample  # this might change with new pytorch dataset structure
+            sample = {"input": sample[0], "target": sample[1]}  # new dataset api will come like this
             self.optimizer.zero_grad()
-
             # perform the preprocess and augmentations in batch
-            img = self.preprocess(source)
-            img = self.augmentations(img)
+            sample = self.preprocess(sample)
+            sample = self.augmentations(sample)
+            sample = self.postprocess(sample)
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/x/trainer.py:103]: Check the script example, we need something  configurable before the model inference, like convert back the labels to long type, etc. Probably with better name

File: examples/train/semantic_segmentation/main.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+def my_app(config: Configuration) -> None:
+
+    # create the model
+    model = nn.Sequential(
+        torchvision.models.segmentation.fcn_resnet50(pretrained=False),
+        Lambda(lambda x: x['out']),
+    )
+
+    def tf(x, y) -> dict:
+        return T.ToTensor()(x), torch.from_numpy(y)
```
Comment (shijianjian) [inline examples/train/semantic_segmentation/main.py:28]: dict?

**Comment by shijianjian** [comment]
BTW, ```pip install kornia[x]``` is not a valid command.

**Comment by edgarriba** [comment]
> BTW, `pip install kornia[x]` is not a valid command.

in the next release it will be available - now you have to run from the root source

**Comment by edgarriba** [comment]
@shijianjian check again - I added `on_before_model` instead plus few other changes

File: examples/train/semantic_segmentation/main.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    valid_dataset = torchvision.datasets.SBDataset(
+        root=to_absolute_path(config.data_path), image_set='val', download=False, transforms=transform)
+
+    # create the dataloaders
+    train_dataloader = torch.utils.data.DataLoader(
+        train_dataset, batch_size=config.batch_size, shuffle=True, num_workers=8, pin_memory=True)
+
+    valid_daloader = torch.utils.data.DataLoader(
+        valid_dataset, batch_size=config.batch_size, shuffle=True, num_workers=8, pin_memory=True)
+
```
Comment (shijianjian) [inline examples/train/semantic_segmentation/main.py:48]: I am thinking if to build it towards a more less general purpose. The boiler code is too much, thinking of intergrating this part to the Trainer directly or so. So dose the preprocess and the postprocess.

File: kornia/x/trainer.py [hunk 6]
```diff
@@ -95,19 +95,19 @@ def fit_epoch(self, epoch: int) -> None:
...
-            img = self.preprocess(source)
-            img = self.augmentations(img)
+            sample = self.preprocess(sample)
+            sample = self.augmentations(sample)
+            sample = self.on_before_model(sample)
             # make the actual inference
-            output = self.model(img)
-            loss = self.criterion(output, target)
+            output = self.model(sample["input"])
+            loss = self.criterion(output, sample["target"])
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/x/trainer.py:121]: in that case in the `ObjectDetectorTrainer` you might need to overrided this part of the code. Besides, could you point what you return ? because to me seems totally wrong return a loss from the model. Another case would be that intermediate features are returned from the model and then we need a special criterion that selects the needed things and computes the actual loss.

File: examples/train/semantic_segmentation/main.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    model_checkpoint = ModelCheckpoint(
+        filepath="./outputs", monitor="iou",
+    )
+
+    trainer = SemanticSegmentationTrainer(
+        model, train_dataloader, valid_daloader, criterion, optimizer, scheduler, config,
+        callbacks={
+            "preprocess": preprocess,
+            "augmentations": augmentations,
+            "on_before_model": on_before_model,
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply examples/train/semantic_segmentation/main.py:106]: I don't thnk so - e.g during the validation if you check you need `prepocess` and `on_before_model` and no `augmentations`. If not augmentations, those names are zero intuitive

File: kornia/x/trainer.py [hunk 8]
```diff
@@ -142,14 +142,17 @@ def fit(self,) -> None:
     def evaluate(self):
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/x/trainer.py:161]: possibly better validate - but definitely no predict. Prediction shouldn't exist since the model.forward might do the job

File: examples/train/semantic_segmentation/main.py (hunk 5 above)
Comment (edgarriba) [reply examples/train/semantic_segmentation/main.py:48]: we can have a lighter layer later but this is plane pytorch in the end. Generalizing datasets is tricky since every problem/user usually have a totally random structure. Maybe a data structure that encapsulates train/valid dataloader - but in the next iteration once we increase the number of supported tasks and we figure out stable code patterns for what we support in terms of models. We don't want to be a general purpose thing

**Review by shijianjian — APPROVED** [review]
